### PR TITLE
Fork choice tests

### PR DESCRIPTION
**Pull request by AgeManning**
## Issue Addressed
#222 
## Proposed Changes
Tested fork-choice rules with YAML testing framework.

### CODE REVIEW

File: eth2/fork_choice/src/optimised_lmd_ghost.rs [hunk 1]
```diff
@@ -212,41 +198,55 @@ where
...
             let mut one_votes = 0;
-            let mut single_candidate = None;
+            let mut single_candidate = (None, false);
 
+            trace!("FORKCHOICE: Child vote length: {}", votes.len());
             for (candidate, votes) in votes.iter() {
-                let candidate_uint = BigEndian::read_u32(candidate);
-                if candidate_uint >> (bit + 1) != bitmask {
+                let candidate_bit: BitVec = BitVec::from_bytes(&candidate);
+                /*
```
Comment (paulhauner) [inline eth2/fork_choice/src/optimised_lmd_ghost.rs]: Just checking you intended to leave these comments in.

File: eth2/fork_choice/src/optimised_lmd_ghost.rs [hunk 2]
```diff
@@ -51,6 +26,9 @@ const EPOCH_LENGTH: u64 = 64;
 // the comparison. Log2_raw takes 2ns according to the documentation.
 #[inline]
 fn log2_int(x: u32) -> u32 {
+    if x == 0 {
+        return 0;
+    }
     log2_raw(x as f32) as u32
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: Yeah. I agree. I guess this should be done at a higher level. I'll find the best place to panic.

File: eth2/fork_choice/src/optimised_lmd_ghost.rs [hunk 3]
```diff
@@ -117,25 +96,31 @@ where
...
+                spec.max_deposit_amount,
+            ) / spec.fork_choice_balance_increment;
             if balance > 0 {
                 if let Some(target) = self.latest_attestation_targets.get(&(index as u64)) {
                     *latest_votes.entry(*target).or_insert_with(|| 0) += balance;
                 }
             }
         }
-
+        trace!("FORKCHOICE: Latest votes: {:?}", latest_votes);
```
[the remark this answers is not in the document]
Comment (AgeManning) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: yeah good call.

File: eth2/fork_choice/src/optimised_lmd_ghost.rs (hunk 1 above)
Comment (AgeManning) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: Yeah.. they are really handy for debugging (figure I'll need them when the other guys disclose a full set of test vectors). I've added and removed them a few times.. so now i'm leaving in with comments.

File: eth2/fork_choice/src/optimised_lmd_ghost.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AgeManning) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: Yep, i'll also put a panic in the slot -> to height. If we are lossy there, I think we should also panic.

**Pull request merged**